### DISCUSSION TITLE
[QUESTION] Terminate a query by cancellation token

### DISCUSSION THREAD

**Discussion by gpgpublickey** [linq2db/linq2db#4472] [General]
I would like to know if async methods querying data in case cancellation token passed to linq2db is triggered will generate query to cancel or will keep running in the db engine under the hood. Thanks

**Comment by jods4** [comment]
Cancellation token is passed to ADO.NET provider methods (such as `ExecuteDataReaderAsync`, etc.).
What happens exactly server-side depends on your specific DB provider, you should check / test.

For example some providers have true async I/O implementation, others just do sync-over-async.
Whether the server-side query will actually be cancelled or not is also an implementation detail of the RDBMS you use. It feels like it should be cancelled but I wouldn't take that as granted for all providers.

**Comment by MaceWindu** [comment]
We have some tests for cancellation and most of providers with async support support it properly at least for tested methods. Only exception is `MySql.Data` provider which never does anything properly from my experience.

**Comment by jods4** [comment]
I also wonder what `Oracle` does, as until the latest release they didn't have true async implementation but performed sync-over-async.

**Comment by MaceWindu** [comment]
For such providers called async method implemented by base ado.net class, which calls sync implementation and checks token cancellation.
E.g. for mentioned above method `ExecuteReaderAsync`

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Data.Common/src/System/Data/Common/DbCommand.cs#L164

**Comment by jods4** [comment]
Notice that the call you linked to only happens before the request actually starts.
Then, during execution the abstract method `DbCommand.Cancel()` is called and what actually happens -- if anything -- is up to provider implementation. [Comment in ado.net source](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Data.Common/src/System/Data/Common/DbCommand.cs#L83-L97) even says: 
>  Cancellation is a suggestion
